src/inference/include/openvino/runtime/async_infer_queue.hpp

```diff
@@ -0,0 +1,197 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+ * @brief A header file that provides wrapper classes for infer requests and callbacks.
+ *
+ * @file openvino/runtime/async_infer_queue.hpp
+ */
+#pragma once
+
+#include <condition_variable>
+#include <map>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include "ie/ie_plugin_config.hpp"
+#include "openvino/runtime/common.hpp"
+#include "openvino/runtime/executable_network.hpp"
+#include "openvino/runtime/infer_request.hpp"
+#include "openvino/runtime/tensor.hpp"
+
+namespace helpers {
+/**
+ * @brief Helper function to get number of jobs based on given network's metrics.
+ *
+ * @param net ExecutableNetwork object.
+ *
+ * @return Returns number of jobs.
+ */
+static size_t num_of_jobs_helper(ov::runtime::ExecutableNetwork& net) {
+    try {
+        auto parameter_value = net.get_metric(METRIC_KEY(SUPPORTED_METRICS));
+        auto supported_metrics = parameter_value.as<std::vector<std::string>>();
+        const std::string key = METRIC_KEY(OPTIMAL_NUMBER_OF_INFER_REQUESTS);
+        if (std::find(supported_metrics.begin(), supported_metrics.end(), key) != supported_metrics.end()) {
+            parameter_value = net.get_metric(key);
+            if (parameter_value.is<unsigned int>())
+                return parameter_value.as<unsigned int>();
+            else
+                throw ov::Exception("Unsupported format for " + key + "! Please specify number of jobs explicitly!");
+        } else {
+            throw ov::Exception("Can't load model: " + key +
+                                " is not supported! Please specify number of jobs explicitly!");
+        }
+    } catch (const std::exception& ex) {
+        throw ov::Exception(ex.what());
+    }
+}
+}  // namespace helpers
+
+namespace ov {
+namespace runtime {
+
+/**
+ * @brief This is an interface of asynchronous inference of requests.
+ *
+ * Based on a concept of jobs pool which holds InferRequests.
+ * Work is automatically dispatched without user synchronization inside one queue.
+ * Advanced workflow can be managed via callbacks and blocking functions like
+ * wait_all() or is_ready().
+ *
+ * @note AsyncInferQueue provides simple and unified interface for multiple asynchronous
+ * calls to OpenVINO's InferRequests.
+ */
+class OPENVINO_RUNTIME_API AsyncInferQueue {
+public:
+    /**
+     * @brief Empty constructor.
+     */
+    AsyncInferQueue();
```

Comment:
If we introduce new API, we need to provide samples or other examples in order to show how user should use this API

Reply:
I have added tests on both C++ and Python side. Demos and other tools will be adjusted to use Queue but we need to merge it first... I won't create a PR to edit all components at once.

Reply:
this is good point, ideally if we offer an impl, the benchamrk_ap and the demos should preferably use that